lucille-core/src/test/java/com/kmwllc/lucille/core/RunnerManagerTest.java

```diff
@@ -1,56 +1,133 @@
-    assertFalse(runnerManager.isRunning());
-    assertTrue(runnerManager.runWithConfig(config));
+    assertFalse(runnerManager.isRunning(runId));
+    assertTrue(!details.hasError());
 
     // Wait for all lucille threads to finish before exiting
-    runnerManager.waitForRunCompletion();
+    runnerManager.waitForRunCompletion(runId);
   }
 
   @Test
-  public void testWaitForRunCompletion() throws InterruptedException, ExecutionException {
+  public void testWaitForRunCompletion() throws Exception {
     RunnerManager runnerManager = RunnerManager.getInstance();
     Config config = ConfigFactory.load("RunnerManagerTest/sleep.conf");
+    String configId = runnerManager.createConfig(config);
+    String runId = Runner.generateRunId();
 
     // Ensure lucille is not running first
-    assertFalse(runnerManager.isRunning());
+    assertFalse(runnerManager.isRunning(runId));
+
+    runnerManager.runWithConfig(runId, configId);
+    long startTime = System.nanoTime();
+
+    log.info("before waiting for run to complete {} ========", runnerManager.isRunning(runId));
+    assertTrue(runnerManager.isRunning(runId));
+
+    runnerManager.waitForRunCompletion(runId);
 
-    runnerManager.runWithConfig(config);
+    long endTime = System.nanoTime();
+    long durationMillis = (endTime - startTime) / 1_000_000;
 
-    // Ensure lucille is running, wait for it stop and ensure its stopped
-    assertTrue(runnerManager.isRunning());
-    runnerManager.waitForRunCompletion();
-    assertFalse(runnerManager.isRunning());
+    log.info("after waiting for run to complete {} ========", runnerManager.isRunning(runId));
+    assertFalse(runnerManager.isRunning(runId));
+
+    log.info("Test execution time: {} ms", durationMillis);
   }
+
+
+  @Test
+  public void testSimultaneousRuns() throws Exception {
+    Config config = ConfigFactory.load("RunnerManagerTest/sleep.conf");
+    RunnerManager runnerManager = RunnerManager.getInstance();
+    String configId = runnerManager.createConfig(config);
+    List<String> runIds = new ArrayList();
+
+    for (int i = 0; i < 5; i++) {
+      String runId = "test-run-" + i;
+      // runnerManager.runWithConfig is non-blocking so we don't need to invoke it here via
+      // CompleteableFuture.runAsync()
+      // but this approach simulates a scenerio where multiple threads are calling it concurrently
+      CompletableFuture.runAsync(() -> {
+        try {
+          runnerManager.runWithConfig(runId, configId);
+          assertFalse(runnerManager.getRunDetails(runId).isDone());
+        } catch (RunnerManagerException e) {
+          throw new RuntimeException(e);
+        }
+      });
+      runIds.add(runId);
+    }
+
+    assertEquals(5, runIds.size());
+
+    StopWatch stopWatch = new StopWatch();
+    stopWatch.start();
+    while (runIds.stream().anyMatch(x -> {
+      RunDetails details = runnerManager.getRunDetails(x);
+      return details == null || !details.isDone();
```

Comment:
again, please do not do a null check here unless you are doing that in an assertion that would fail the test
otherwise we want a nullpointer exception to bubble up and fail the test
we should never see a null here and if we do it's an error that needs to be surfaced

Reply:
probably kruft from debugging ... removed